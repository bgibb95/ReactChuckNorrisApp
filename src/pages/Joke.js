import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import fetchJokeFromCategory from '../actions/joke';
import {
  RightContainer,
  LeftContainer,
  BasicLink,
  MainTitle,
  NewCategoryLink,
  Page,
} from '../styles/common';
import {
  Row,
  PrimaryButton,
  JokeIcon,
  JokeHeader,
} from '../styles/joke';

export class Joke extends Component {
  constructor(props) {
    super(props);
    this.getJoke = this.getJoke.bind(this);
  }

  componentWillMount() {
    this.getJoke();
  }

  getJoke() {
    const { fetchJokeFromCategory: getJokeFromCategory, match } = this.props;
    const { category } = match.params;
    getJokeFromCategory(category);
  }

  render() {
    const { joke, loading } = this.props;
    return (
      <Page>
        <LeftContainer>
          <MainTitle>
            {joke.category}
            {' '}
            Joke
          </MainTitle>
          <Row>
            <PrimaryButton type="button" onClick={this.getJoke}> New Joke</PrimaryButton>
            <NewCategoryLink to="/">New Category</NewCategoryLink>
          </Row>
        </LeftContainer>
        <RightContainer loading={loading}>
          { joke.icon_url ? <JokeIcon src={`${joke.icon_url}`} alt="jokeImage" /> : null}
          <JokeHeader>{joke.value}</JokeHeader>
          <BasicLink href={`${joke.url}`} rel="noopener noreferrer" target="_blank">See joke</BasicLink>
        </RightContainer>
      </Page>
    );
  }
}

Joke.propTypes = {
  match: PropTypes.shape({}).isRequired,
  fetchJokeFromCategory: PropTypes.func.isRequired,
  joke: PropTypes.shape({}).isRequired,
  loading: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  joke: state.joke.joke,
  loading: state.joke.loading,
});

export default connect(mapStateToProps, { fetchJokeFromCategory })(Joke);
