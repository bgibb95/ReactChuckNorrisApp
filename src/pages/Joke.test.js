import React from 'react';
import { create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Joke } from './Joke';

Enzyme.configure({ adapter: new Adapter() });

describe('Joke page', () => {
  const props = {
    match: { params: { category: 'dev' } },
    fetchJokeFromCategory: jest.fn(),
    joke: {},
    loading: false,
  };

  const component = create(<MemoryRouter><Joke {...props} /></MemoryRouter>);

  test('It matches the snapshot', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('Should call getJoke on mount', () => {
    expect(props.fetchJokeFromCategory).toHaveBeenCalled();
  });

  test('Should call getJoke on button click', () => {
    const wrapper = mount(<MemoryRouter><Joke {...props} /></MemoryRouter>);
    wrapper.find('button').at(0).simulate('click');
    expect(props.fetchJokeFromCategory).toHaveBeenCalledTimes(3);
  });
});
