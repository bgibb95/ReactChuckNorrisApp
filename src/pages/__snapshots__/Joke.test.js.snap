// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joke page It matches the snapshot 1`] = `
<div
  className="common__Page-sc-4zl79x-0 tfEnI"
>
  <div
    className="common__Container-sc-4zl79x-6 common__LeftContainer-sc-4zl79x-8 UmaQz"
  >
    <h1
      className="common__MainTitle-sc-4zl79x-5 WItnH"
    >
       
      Joke
    </h1>
    <div
      className="joke__Row-a9hi5-3 gPDRKs"
    >
      <button
        className="joke__PrimaryButton-a9hi5-2 eIfJhA"
        onClick={[Function]}
        type="button"
      >
         New Joke
      </button>
      <a
        className="common__StyledLink-sc-4zl79x-3 common__NewCategoryLink-sc-4zl79x-4 eQfvQT"
        href="/"
        onClick={[Function]}
      >
        New Category
      </a>
    </div>
  </div>
  <div
    className="common__Container-sc-4zl79x-6 common__RightContainer-sc-4zl79x-7 hVahFF"
  >
    <h3
      className="joke__JokeHeader-a9hi5-1 vYAMl"
    />
    <a
      className="common__BasicLink-sc-4zl79x-2 eEPEnA"
      href="undefined"
      rel="noopener noreferrer"
      target="_blank"
    >
      See joke
    </a>
  </div>
</div>
`;
